### PR TITLE
fix(ConditionBuilder): can't change month or year for Date type

### LINKED ISSUE #5517: [Condition Builder]: Can't change month or year for Date type.
### Package

Carbon for IBM Products

### Description

When using condition builder date user cannot change month or year. The calendar popup closes.

### Component(s) impacted

ConditionBuilder

### Browser

_No response_

### @carbon/ibm-products (previously @carbon/ibm-cloud-cognitive) version

2.39

### Severity

Severity 1 = The design is broken in a critical way that blocks users from completing tasks or damages the brand. Affects major functionality, no workaround.

### Product/offering

FTM

### CodeSandbox or Stackblitz example

https://ibm-products.carbondesignsystem.com/?path=/story/ibm-products-components-conditionbuilder--condition-builder-with-initial-state&globals=viewport:basic

### Steps to reproduce the issue (if applicable)

Go https://ibm-products.carbondesignsystem.com/?path=/story/ibm-products-components-conditionbuilder--condition-builder-with-initial-state&globals=viewport:basic

1.Press Add Condition
2. Change Region to Date
3. Select is berween
4. Click on start date
5. Try to change a month.

### Release date (if applicable)

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/ibm-products/blob/main/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/ibm-products/issues) for duplicate issues

### PR DESCRIPTION
**Pull request by amal-k-joy**
Closes #5517 

Can't change month or year for Date type

#### What did you change?
As the date-picker element is outside the popover, and when we click on anywhere on date picker , popover is getting closed. As the fix, popover closing is prevented if the click is from date picker.

#### How did you test and verify your work?
Local storybook

### CODE REVIEW

**Comment by matthewgallo** [comment]
You might be able to just use the `appendTo` prop here
https://react.carbondesignsystem.com/?path=/docs/components-datepicker--overview#datepicker-appendto

File: packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js [hunk 1]
```diff
@@ -85,14 +87,28 @@ export const ConditionBuilderItem = ({
...
+
   return (
     <Popover
       open={open}
       isTabTip
       role="gridcell"
       className={popOverClassName}
       onRequestClose={() => {
-        setOpen(false);
+        if (eligibleToClose()) {
```
Comment (davidmenendez) [inline packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js:109]: lets not add logic like this in the HTML. i would create an onClose handler that deals with all this.

File: packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js [hunk 2]
```diff
@@ -85,14 +87,28 @@ export const ConditionBuilderItem = ({
     }
   }, [contentRef, open]);
 
+  const eligibleToClose = () => {
+    if (
+      event?.target?.closest?.(`.${carbonPrefix}--date-picker__calendar`) ||
```
Comment (davidmenendez) [inline packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js:92]: missing the `event` in the function definition

**Review by davidmenendez — CHANGES_REQUESTED** [review]
i would definitely check and see if @matthewgallo suggestion works and if it doesn't i have some additional comments.

also i don't know if this is related, but i noticed a bug when you open calendar and then click out you end up with `nan` in the value

Screenshot 2024-06-28 at 9 35 06 AM

**Comment by davidmenendez** [comment]
there's also a weird bug going on where if you open the calendar and then click on the calendar labels (`start` `end`) it does this strange focus open close behavior

https://github.com/carbon-design-system/ibm-products/assets/6370760/5dd62257-6b7f-4853-9c4a-1e491db04b5c

**Comment by amal-k-joy** [comment]
> You might be able to just use the `appendTo` prop here https://react.carbondesignsystem.com/?path=/docs/components-datepicker--overview#datepicker-appendto

Thanks @matthewgallo 
This worked.

File: packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js (hunk 1 above)
Comment (amal-k-joy) [reply packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js]: resolved

File: packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js (hunk 2 above)
Comment (amal-k-joy) [reply packages/ibm-products/src/components/ConditionBuilder/ConditionBuilderItem/ConditionBuilderItem.js]: Carbon is not providing the `event` for `onRequestClose` callback. But i refined the approach based on  Matt suggested.

**Comment by amal-k-joy** [comment]
> i would definitely check and see if @matthewgallo suggestion works and if it doesn't i have some additional comments.
> 
> also i don't know if this is related, but i noticed a bug when you open calendar and then click out you end up with `nan` in the value
> 
> Screenshot 2024-06-28 at 9 35 06 AM

As Matt suggested I have passed the popover `ref` to render the datapicker, so that it wont get closed when clicked. Also when an invalid date is selected , it will show the invalid date text now instead on `NaN`

**Review by davidmenendez — APPROVED** [review]

**Review by makafsal — APPROVED** [review]

**Pull request merged**